stoppable_workers.go

```diff
@@ -0,0 +1,64 @@
+package utils
+
+import (
+	"context"
+	"errors"
+	"sync"
+)
+
+// StoppableWorkersAlreadyStopped is returned by Add when the StoppableWorkers
+// instance has already been stopped.
+var StoppableWorkersAlreadyStopped = errors.New("cannot add worker: already stopped")
+
+// StoppableWorkers is a collection of goroutines that can be stopped at a
+// later time.
+type StoppableWorkers struct {
+	mu         sync.RWMutex
+	ctx        context.Context
+	cancelFunc func()
+
+	workers sync.WaitGroup
+}
+
+// NewStoppableWorkers creates a new StoppableWorkers instance. The instance's
+// context will be derived from passed in context.
+func NewStoppableWorkers(ctx context.Context) *StoppableWorkers {
```

Comment:
I think `NewStoppableWorkers` should not always derive from `context.Background()` (hence the context parameter here). See the `nested workers` sub test I wrote; NetCode has objects that manage workers, and some of those workers _have their own_ workers. We want to allow establishing a context hierarchy such that a `StoppableWorkers` instance can be `Stop`ped and, in doing so, stop the workers of any of its workers `StoppableWorkers` instances.